shared_controller_node:
  ros__parameters:
    controller_frequency: 20.          #in second. Amount of time needed to complete one cycle of vel calculation
    joystick_noise: false
    max_linear_vel: 0.3      # max linear velocity in x direction (0.3-1.6)(m/s) 
    min_linear_vel: -0.3      # min linear velocity. 
                               # Can be negative in case of backward movement . (-0.3 - -0.8)(m/s)
                               # (Maximum backward vel)
    max_yaw_rate: 0.7          # Maximum yaw rate in both clockwise and counter clockwise direction. Can only be positive. Min yaw rate = 0 (rad/s)
    max_acceleration: 1.2       # max acceleration in both direction (forward backward) ( 0.28  - 1.38m/ss)
    max_decceleration: 1.2    # max decceleration in both direction (forward backward) (1.2 - 2.5m/ss)
    max_yaw_acceleration: 2.5  # maximum yaw rate acceleration in both clockwise and counter clockwise (rad/ss)


    shared_mppi_controller:
      time_steps: 60
      model_dt: 0.05
      batch_size: 2000
      vx_std: 0.3
      vy_std: 0.3
      wz_std: 0.4
      vx_max: 0.3
      vx_min: -0.3
      vy_max: 0.
      wz_max: 0.7
      ax_max: 1.2  
      ax_min: -1.2
      ay_max: 0.
      az_max: 2.5 
      iteration_count: 1
      prune_distance: 1.7
      temperature: 0.3
      gamma: 0.015
      motion_model: "DiffDrive"
      visualize: true
      reset_period: 1.0 # (only in Humble)
      retry_attempt_limit: 3
      regenerate_noises: false
      TrajectoryVisualizer:
        trajectory_step: 1000
        time_step: 100
      critics: ["ConstraintCritic", "CostCritic", "ObstaclesCritic", "GoalCritic", "GoalAngleCritic", "PreferForwardCritic", "UserSpeedCritic"]
      ConstraintCritic:
        enabled: true
        cost_power: 1
        cost_weight: 4.0
      GoalCritic:
        enabled: true
        cost_power: 2
        cost_weight: 4.0
        threshold_to_consider: 3.0
      GoalAngleCritic:
        enabled: true
        cost_power: 1
        cost_weight: 3.0
        threshold_to_consider: 0.5
      PreferForwardCritic:
        enabled: false
        cost_power: 1
        cost_weight: 5.0
        threshold_to_consider: 0.5
      ObstaclesCritic:
        enabled: true
        cost_power: 1
        repulsion_weight: 3.0
        critical_weight: 20.0
        consider_footprint: true
        collision_cost: 10000.0
        collision_margin_distance: 0.1
        near_goal_distance: 0.1
        inflation_radius: 0.55 # (only in Humble)
        cost_scaling_factor: 10.0 # (only in Humble)
      CostCritic:
        enabled: false
        cost_power: 2
        cost_weight:  3.81
        critical_cost: 300.0
        consider_footprint: true
        collision_cost: 1000000.0
        near_goal_distance: 0.5
        trajectory_point_step: 4
      UserSpeedCritic:
        enable: true
        cost_power: 1
        cost_weight: 4.0
        decay_rate: 0.2

local_costmap/local_costmap:
  ros__parameters: 
    update_frequency: 20.0
    publish_frequency: 20.0
    footprint: "[[0.45, 0.32],[0.45,-0.32],[-0.45,-0.32],[-0.45,0.32]]"  # Robot footprint
    global_frame: base_footprint
    robot_base_frame: base_footprint
    use_sim_time: False
    rolling_window: True  # Enable rolling window for the local costmap
    width: 6  # Size of the local costmap (around the robot)
    height: 6
    resolution: 0.06
    plugins: ["obstacle_layer", "inflation_layer"]  # Add desired layers
    obstacle_layer:
      plugin: "nav2_costmap_2d::ObstacleLayer"
      enabled: true
      footprint_clearing_enabled: false
      observation_sources: scan  # Subscribe to scan data
      footprint_clearing_enabled: true
      max_obstacle_height: 2.0
      combination_method: 0
      scan:
        topic: /scan
        sensor_frame: base_footprint
        obstacle_max_range: 3.0
        obstacle_min_range: 0.0
        raytrace_max_range: 5.0
        raytrace_min_range: 0.0
        max_obstacle_height: 2.0
        min_obstacle_height: 0.0
        clearing: true
        marking: true
        data_type: "LaserScan"
    inflation_layer:
      plugin: "nav2_costmap_2d::InflationLayer"
      enabled: true
      inflation_radius: 0.55
      cost_scaling_factor: 1.0
      inflate_unknown: false
      inflate_around_unknown: true
